signal: handle shutdown properly

hsjoberg opened the pull request:
This PR addresses issues related to the mobile lnd builds, where calling `stopDaemon` and subsequently calling LndMobile's `Start` results in panic.

This is a part of an on-going work to address mobile restarting problems #4492.

halseth commented on signal/signal.go:
```diff
@@ -93,6 +98,7 @@ func mainInterruptHandler() {
 		case <-quit:
 			log.Infof("Gracefully shutting down.")
 			close(shutdownChannel)
+			atomic.StoreInt32(&started, 0)
```
could add this to a defer in the beginning of the method.

hsjoberg replied:
Oh yeah that definitely makes sense.

halseth commented on signal/signal.go:
```diff
@@ -15,21 +15,21 @@ import (
 var (
 	// interruptChannel is used to receive SIGINT (Ctrl+C) signals.
-	interruptChannel = make(chan os.Signal, 1)
+	interruptChannel chan os.Signal
```
could we instead encapsulate these in a struct?

so it could be used like
```go
signal := signal.Intercept()
<-signal.ShutdownChan()

```

fiatjaf replied:
Why does this matter? In my understanding a simple encapsulation like this won't change the fact that these channels will still live as globals under `signal`. It will just require some unnecessary changes on lnd entry points and this confusing `signal := signal` assignment.

Or shouldn't the channels live as globals under `signal`? (That would require relatively big channels on that package.)

carlaKC replied:
I think it would be possible to create a single `Interceptor` struct which contains all of the channels that we use in the signal package and then define the various functions that currently reference the globals (eg `Alive`/`Listening`)  as functions on that struct? 

The downside there is that we need to pass around a single `interceptor` struct to the places where we currently call these functions, but it doesn't look too bad? The only tricky place I can see is in `DefaultConfig` because we haven't called `Intercept` yet, but we can just switch in a non-nil interceptor when we `LoadConfig` - happy to help out with this part in more detail, because I added the change which needs the interceptor for our logger!

halseth replied:
Yeah, my point was to remove them from the global scope entirely :)

cfromknecht replied:
+1 for moving to contained struct. this will also be useful in the future if we want to run multiple lnd's in the same process, as otherwise they would all share the same global signals.

halseth reviewed (changes requested):
Thanks for the PR! Definitely needed for some of the mobile applications :)

hsjoberg commented:
Okay, I've addressed the points.
Added new commits, will squash once we're ready.

There are still global variables and functions because of the way `signal.go` works, I'm not sure if it makes sense for a complete overhaul here and perhaps someone more experienced in Go and lnd's codebase should address that.

carlaKC commented on lnd.go:
```diff
@@ -187,7 +187,7 @@ type rpcListeners func() ([]*ListenerWithSignal, func(), error)
 // This function starts all main system components then blocks until a signal
 // is received on the shutdownChan at which point everything is shut down again.
-func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan <-chan struct{}) error {
+func Main(cfg *Config, lisCfg ListenerCfg, shutdownChan chan struct{}) error {
```
`shutdownChan` can still be `<- chan struct` if you update `waitForWalletPassword` to also take a receive only channel?

hsjoberg replied:
Thanks! Will fix

carlaKC reviewed (commented):
Thanks for the PR @hsjoberg! 

I think that it would be worth trying to refactor this into an `Interceptor` struct if possible and get rid of the globals entirely? I don't think the change will be too much bigger, just a bit of gymnastics around `DefaultConfig`.

halseth commented:
> There are still global variables and functions because of the way `signal.go` works

Why do they have to stay globals exactly? It seems like you already moved `InterruptChannel` out of the global scope, is something in the way for doing this for the remaining vars?

hsjoberg commented:
@carlaKC Thank you for your review Carla. I will work on making it non-global.

@halseth I don't think is any specific reason that they have to stay global, it's mostly my lack of understanding of the codebase.

hsjoberg commented:
I'm not sure about the global `started` variable though. The reason for its existence is to make sure that `signal.Intercept` is only called once.
https://github.com/lightningnetwork/lnd/blob/cd41cd4d2d500f2d2b39e56d20932c7ec657b103/signal/signal.go#L37-L40

hsjoberg commented:
Rebased

carlaKC commented on build/log.go:
```diff
@@ -50,8 +51,8 @@ type LogWriter struct {
 // writer is shared amongst all instantiations.
-func NewSubLogger(subsystem string,
-	genSubLogger func(string) btclog.Logger) btclog.Logger {
+func NewSubLogger(subsystem string, interceptor signal.Interceptor,
```
Can we get away with not needing to pass the interceptor in here and leave that the the `genSubLogger` function? That way we don't need to pass all the empty interceptors in. 

So the signature here is: 
`NewSubLogger(subsystem string, genSubLogger func(string, func()) btclog.Logger)`

When we don't have access to an interceptor we can just call it with:
`build.NewSublogger(subsystem, nil)`

When we do have access to an interceptor, we use:
```
shutdown:=func(){
  if !interceptor.Listening() {
    return
  }

  interceptor.RequestShutdown()
}

build.NewSublogger(subsystem, shutdown)
```

Tried this out quickly [here](https://github.com/carlaKC/lnd/commit/abfce52064fe5d1e684e81d5a2f7e64fb0ffb56c) and it seems to work, although the code could probably be a bit cleaner 😅

hsjoberg replied:
Thanks, I cherry-picked your commit. It is much better.

carlaKC commented on cmd/lnd/main.go:
```diff
@@ -10,9 +10,11 @@ import (
 
 func main() {
+	// Hook interceptor for os signals.
+	signal, err := signal.Intercept()
```
Need to handle this error.

hsjoberg replied:
Fixed.

carlaKC reviewed (changes requested):
This is starting to look really good! My only major comment is that I think we can simplify `build.NewSublogger` a bit, have added a detailed comment in my review. 

A few of the PR checks have failed (eg. linter and making sure each commit compiles), so also just have a look at those? Squashing some of the fixup commits should help.

halseth commented on autopilot/log.go:
```diff
@@ -12,7 +13,7 @@ var log btclog.Logger
 // The default amount of logging is none.
 func init() {
-	UseLogger(build.NewSubLogger("ATPL", nil))
+	UseLogger(build.NewSubLogger("ATPL", signal.Interceptor{}, nil))
```
this change seems a bit large and unnecessary to me. I get that it is a way to request a shutdown without going through a global var, but I'm hesitant to merge such a large change.

I think it is better to leave the global connection for now, until we have a better understanding of whether this is needed. A way would just keep the latest created struct in the global scope such that it can easily be requested to shut down.

hsjoberg replied:
@halseth I think the scope has been reduced significantly by @carlaKC's [code improvements](https://github.com/carlaKC/lnd/commit/abfce52064fe5d1e684e81d5a2f7e64fb0ffb56c). So perhaps we can continue with the current path.

hsjoberg commented on rpcserver.go:
```diff
@@ -5530,8 +5529,7 @@ func (r *rpcServer) GetNetworkInfo(ctx context.Context,
 	_ *lnrpc.StopRequest) (*lnrpc.StopResponse, error) {
-
-	signal.RequestShutdown()
+	// signal.RequestShutdown()
```
Oops, fixed.

hsjoberg commented on signal/signal.go:
```diff
@@ -14,49 +14,62 @@ import (
 
-	// quit is closed when instructing the main interrupt handler to exit.
-	quit = make(chan struct{})
+type Interceptor struct {
```
Added, although not perfect.

hsjoberg commented:
I think I've addressed all points. Please Take A Look again.

I've tested @ 2b369f04cf074921a65290035b6e6601452b90a0, both normal and mobile builds work fine.

hsjoberg commented:
I'll squash once you've had the time to check the new changes.

halseth commented on build/log_shutdown.go:
```diff
@@ -38,16 +39,3 @@ func (s *ShutdownLogger) Critical(v ...interface{}) {
-		return
-	}
-
-	s.Logger.Info("Sending request for shutdown")
```
Nice, this was a clever solution :)

Can move this log statement to where `shutdown` is called perhaps?

hsjoberg replied:
Yes much better, @carlaKC takes full credit for this one. :)

I added back the log message to Critical and Criticalf. I guess that makes sense?

https://github.com/lightningnetwork/lnd/blob/4d45f50f1eb7a76b812c99aacf61b72f1c5ee7f2/build/log_shutdown.go#L28-L32

https://github.com/lightningnetwork/lnd/blob/4d45f50f1eb7a76b812c99aacf61b72f1c5ee7f2/build/log_shutdown.go#L39-L43

halseth commented on signal/signal.go:
```diff
@@ -76,30 +91,31 @@ func mainInterruptHandler() {
 // Listening returns true if the main interrupt handler has been started, and
 // has not been killed.
-func Listening() bool {
+func (channels Interceptor) Listening() bool {
```
style suggestion `func (c *Interceptor)...`

hsjoberg replied:
Thanks, fixed.

halseth commented on signal/signal.go:
```diff
@@ -14,49 +14,64 @@ import (
 // callback registration.
 // It must be run as a goroutine.
-func mainInterruptHandler() {
+func mainInterruptHandler(channels Interceptor) {
```
`func (c *Intereptor) mainInterruptHandler()`

hsjoberg replied:
Fixed.

halseth commented on signal/signal.go:
```diff
@@ -14,49 +14,64 @@ import (
-	shutdownChannel = make(chan struct{})
-)
+	// ShutdownChannel is closed once the main interrupt handler exits.
+	ShutdownChannel chan struct{}
```
I think this is the only channel that needs to be exported here?

halseth replied:
We can even keep the `ShutdownChannel()` function and keep everything in this struct private.

hsjoberg replied:
Definitely! Fixed.

halseth reviewed (changes requested):
Ok, I'm more onboard with these changes now :)

Would be nice to not have to pass the interceptor to `ValidateConfig` but that doesn't seem straight forward the way that method is now initializing the loggers.

carlaKC commented on mobile/bindings.go:
```diff
@@ -82,8 +83,7 @@ func Start(extraArgs string, unlockerReady, rpcReady Callback) {
 	go func() {
-		if err := lnd.Main(
-			loadedConfig, cfg, signal.ShutdownChannel(),
+		if err := lnd.Main(loadedConfig, cfg, interceptor,
```
nit: leave these params on their own newline, or flatten into a single line that ends at 80

hsjoberg replied:
Thanks, I changed them back to a new line.

carlaKC commented on signal/signal.go:
```diff
@@ -14,49 +14,64 @@ import (
+		Quit:                   make(chan struct{}),
 	}
 
+	// shutdownRequestChannel = make(chan struct{})
```
Can delete these lines?

hsjoberg replied:
Oops, definitely!

carlaKC reviewed (approved):
LGTM pending some of the changes @halseth suggested. 

Thanks for sticking with this one @hsjoberg!

hsjoberg commented:
@halseth @carlaKC Please Take A Look again.

halseth commented on cmd/lncli/cmd_open_channel.go:
```diff
@@ -424,7 +424,8 @@ func openChannelPsbt(ctx *cli.Context, client lnrpc.LightningClient,
 	}
 
-	if err := signal.Intercept(); err != nil {
+	signal, err := signal.Intercept()
```
To avoid colliding with the package name, perhaps name it `shutdownInterceptor`?

hsjoberg replied:
Fixed.

halseth commented on cmd/lnd/main.go:
```diff
@@ -10,9 +10,16 @@ import (
 
 func main() {
+	// Hook interceptor for os signals.
+	signal, err := signal.Intercept()
```
same re: trying to avoid colliding with the package name

hsjoberg replied:
Fixed.

halseth commented on signal/signal.go:
```diff
@@ -45,18 +56,19 @@ func Intercept() error {
 // It must be run as a goroutine.
-func mainInterruptHandler() {
+func (c *Interceptor) mainInterruptHandler() {
+	defer atomic.StoreInt32(&started, 0)
```
to we still need this global var?

hsjoberg replied:
I am not sure, right now `signal.Intercept` checks the value of it, as to prevent another signal interception.

https://github.com/lightningnetwork/lnd/blob/68c07352eb3279b9b91292509d369c6f43587b7a/signal/signal.go#L41-L44

carlaKC replied:
I think this can remain as is, can't see a case for multiple interceptors in the same process? And we would need to change some of the other functions, so perhaps in a followup if we really feel the need to kill all the globals.

hsjoberg replied:
Yes, I suggest we leave the started global as is, removing or changing it seems to open up more questions than would close at the moment.
It would be the last thing to fix to make the signal package completely "non-global".

halseth reviewed (commented):
Ok, starts to look good! Can you squash it down and I will take a final look :)

hsjoberg commented:
> Would be nice to not have to pass the interceptor to ValidateConfig but that doesn't seem straight forward the way that method is now initializing the loggers.

I agree, it would be better to not have to pass it to `ValidateConfig`, but not sure about this one.

hsjoberg commented:
Squashed and rebased.

carlaKC reviewed (approved):
I think this can go in as is, just some formatting nits left.

hsjoberg commented on config.go:
```diff
@@ -569,7 +570,7 @@ func LoadConfig() (*Config, error) {
 // illegal values or combination of values are set. All file system paths are
 // normalized. The cleaned up config is returned on success.
-func ValidateConfig(cfg Config, usageMessage string) (*Config, error) {
+func ValidateConfig(cfg Config, usageMessage string, interceptor signal.Interceptor) (*Config, error) {
```
Fixed.

hsjoberg commented:
Yes, I suggest we leave the `started` global as is, removing or changing it seems to open up more questions than would close at the moment.
It would be the last thing to fix to make the signal package completely "non-global".

hsjoberg commented:
Restarting has been tested to work @ c820edb688e84df387a6973d849ece80cc45f7dd together with the other code changes required: https://github.com/lightningnetwork/lnd/issues/4492

halseth reviewed (changes requested):
Awesome, this turned out very nice IMO 👍 

Just a few comments about adding some more documentation, otherwise this LGTM 🏁

hsjoberg commented on build/logrotator.go:
```diff
@@ -14,10 +14,6 @@ import (
 type RotatingLogWriter struct {
-	// GenSubLogger is a function that returns a new logger for a subsystem
-	// belonging to the current RotatingLogWriter.
-	GenSubLogger func(string) btclog.Logger
```
This is much appreciated help from @carlaKC, so she takes full cred here.

hsjoberg commented on log.go:
```diff
@@ -87,12 +87,34 @@ var (
 	atplLog = addLndPkgLogger("ATPL")
 )
 
+func genSubLogger(root *build.RotatingLogWriter,
```
@halseth Added. Although as noted I did not write with this code, so I am not very familiar with it. Trying my best to make an intelligible comment.

hsjoberg commented on log.go:
```diff
@@ -101,51 +123,53 @@ func SetupLoggers(root *build.RotatingLogWriter) {
-	useLoggers ...func(btclog.Logger)) {
+	interceptor signal.Interceptor, useLoggers ...func(btclog.Logger)) {
+
+	genLogger := genSubLogger(root, interceptor)
```
Added.

hsjoberg commented on log.go:
```diff
@@ -87,12 +87,34 @@ var (
+	}
+
+	// Return a function which will create a sublogger from our root
+	// logger with out shutdown fn.
```
Fixed

hsjoberg commented on build/logrotator.go:
```diff
@@ -39,16 +35,18 @@ func NewRotatingLogWriter() *RotatingLogWriter {
 	}
 }
 
+// GenSubLogger creates a new sublogger
```
Extended text.

hsjoberg commented:
@halseth Please Take A Look again.

As the new RPC server changes has landed in master I have rebased against it.
I need to test so restarting lnd still works though, but I think I managed to get it right.

hsjoberg commented:
Okay, I've verified that this code works on top of #5010.

However, there seems to be some leftover regarding the walletunlocker that will give an error once the RPC server is closing. Note though that restarting is still fine and works.

I don't recall seeing this before. Maybe the regression was introduced recently?

```
2021-03-15 16:21:50.305 2938-3625/com.example.lndtest I/GoLog: 2021-03-15 15:21:50.304 [INF] RPCS: Stopping RPC Server
[...]
2021-03-15 16:21:50.307 2938-3605/com.example.lndtest I/GoLog: 2021-03-15 15:21:50.307 [ERR] RPCS: [/lnrpc.WalletUnlocker/UnlockWallet]: got no unlock message before timeout
2021-03-15 16:21:50.307 2938-3607/com.example.lndtest I/GoLog: 2021-03-15 15:21:50.307 [INF] BTCN: Canceling block subscription: id=1
2021-03-15 16:21:50.308 2938-3605/com.example.lndtest E/MainActivity: onError unlockWallet
2021-03-15 16:21:50.308 2938-3605/com.example.lndtest W/System.err: go.Universe$proxyerror: rpc error: code = Unavailable desc = transport is closing
[...]
2021-03-15 16:21:50.414 2938-3607/com.example.lndtest I/GoLog: 2021-03-15 15:21:50.414 [INF] LTND: Shutdown complete
```
(omitted irrelevant log messages)

halseth commented:
@hsjoberg I think that happens if the wallet is still waiting to be unlocked when lnd gets shut down. I don't think it is new, maybe just a consequence of that this was not possible to do before?

Do you think it is a problem with lnd, or can the error be handled gracefully from the app? (since the app is shutting down lnd, it should expect an error if it has an ongoing unlock call).

hsjoberg commented:
> I think that happens if the wallet is still waiting to be unlocked when lnd gets shut down. I don't think it is new, maybe just a consequence of that this was not possible to do before?

@halseth Well, this is after the wallet has been unlocked.

But what you said appears to be true, for some reason a response from unlockWallet is never sent.
https://github.com/lightningnetwork/lnd/blob/a321dac1d83459b5cc1590d9e5e721d92e1912cf/walletunlocker/service.go#L477-L478
So when the RPC error goes down, an error will be sent.

> Do you think it is a problem with lnd, or can the error be handled gracefully from the app? (since the app is shutting down lnd, it should expect an error if it has an ongoing unlock call).

Oh yeah this can be dealt with app-wise. But it would be nice if we could fix this.

halseth commented:
> > I think that happens if the wallet is still waiting to be unlocked when lnd gets shut down. I don't think it is new, maybe just a consequence of that this was not possible to do before?
> 
> @halseth Well, this is after the wallet has been unlocked.
> 
> But what you said appears to be true, for some reason a response from unlockWallet is never sent.
> https://github.com/lightningnetwork/lnd/blob/a321dac1d83459b5cc1590d9e5e721d92e1912cf/walletunlocker/service.go#L477-L478
> 
> 
> So when the RPC error goes down, an error will be sent. This is definitely a regression.
> > Do you think it is a problem with lnd, or can the error be handled gracefully from the app? (since the app is shutting down lnd, it should expect an error if it has an ongoing unlock call).
> 
> Oh yeah this can be dealt with app-wise. But it would be nice if we could fix this.

What does your config look like? (interested in `no-seed-backup` and `no-nacaroons` in particular)

hsjoberg commented:
@halseth This is the config (but with the neutrino node changed):
https://github.com/hsjoberg/lndmobile-speed-tests/blob/0ccd831762deb347b1a770c3f066e87337dbfddb/app/src/main/java/com/example/lndtest/MainActivity.java#L40-L67

So `no-macaroons` is used here actually. Seem to be a leftover. I'll try without it.

hsjoberg commented:
Oh yeah everything works flawlessly without `no-macaroons`!

halseth commented:
> Oh yeah everything works flawlessly without `no-macaroons`!

OK, awesome. What happens is that it will wait for the macaroon that never comes, and return that error on shutdown. The reason we didn't see this before I suspect, is that we would shut down the wallet unlocker (pre #4985) after unlock, such that we wouldn't get into this situation.

Not a big deal, but I can put up a PR to fix it 😄

hsjoberg commented:
> OK, awesome. What happens is that it will wait for the macaroon that never comes, and return that error on shutdown. The reason we didn't see this before I suspect, is that we would shut down the wallet unlocker (pre #4985) after unlock, such that we wouldn't get into this situation.

Right! Thanks for the explanation.

> Not a big deal, but I can put up a PR to fix it 😄

Yeah I guess it's nice to fix it eventually. 🙂 But I agree, not a big deal and not an issue for me personally.

> EDIT: for reference, this is the line that we would call when shutting down the wallet unlocker, which is no longer done: #4985 (files)

Aah got it!

---

Anyway regarding this pull request; it has been squashed and rebased against current master and does work as intended.

halseth commented:
Here is the fix you can try: https://github.com/lightningnetwork/lnd/pull/5112



> Anyway regarding this pull request; it has been squashed and rebased against current master and does work as intended.

Cool, will go back to reviewing it. Anecdote finished 😅

halseth commented on signal/signal.go:
```diff
@@ -14,29 +14,40 @@ import (
 // Intercept starts the interception of interrupt signals. Note that this
 // function can only be called once.
-func Intercept() error {
+func Intercept() (Interceptor, error) {
```
nit: godoc above is now slightly outdated. Should say something like "Note that any previous active interceptor must be stopped before a new one can be created".

hsjoberg replied:
Fixed.

hsjoberg commented on signal/signal.go:
```diff
@@ -76,30 +88,31 @@ func mainInterruptHandler() {
 			log.Infof("Gracefully shutting down.")
-			close(shutdownChannel)
+			close(c.shutdownChannel)
+			signal.Stop(c.interruptChannel)
```
This is to make sure that the signal package [stops](https://pkg.go.dev/os/signal?utm_source=gopls#Stop) sending signals to the interrupt channel. The `interruptChannel` will be re-allocated in `Intercept()` in the next run.

Perhaps it's superfluous, but seems appropriate to call stop.

halseth replied:
I agree, good catch! It was probably a bug that this wasn't done earlier tbh

halseth commented:
Linter is failing:
```
cmd/lncli/commands.go:46:12: ShutdownChannel not declared by package signal (typecheck)
		<-signal.ShutdownChannel()
```

halseth commented:
@hsjoberg I think there was code recently merged to `lncli` that references the signal package, so if you rebase you should se where it fails.

hsjoberg commented:
@halseth Ah gotcha. Will rebase.

hsjoberg commented:
Okay I have based at least.

But sigh... I don't know right now how to solve this one.

hsjoberg commented:
There's no place where `signal.Intercept()` is called in the lncli package (with one exception for openChannelPsbt specifically) so I am unsure this code is even supposed to work.

halseth commented:
> There's no place where `signal.Intercept()` is called in the lncli package (with one exception for openChannelPsbt specifically) so I am unsure how this code is even supposed to work.

Hm, I think you might be right. Maybe it doesn't work really 😛

halseth commented:
Can be rebased now that #5119 is in!

hsjoberg commented:
Rebased and fixed the lncli package -- although I am unsure about the best approach here. Previously `openChannelPsbt()` started the signal interception, but now with that moved to `getContext()`, we need a way to pass the Interceptor instance to `openChannelPsbt()`.
So I see two approacher here, either move the signal interception out of `getContext()` or make `getContext()` return the `Interceptor` instance.

halseth commented:
> Rebased and fixed the lncli package -- although I am unsure about the best approach here. Previously `openChannelPsbt()` started the signal interception, but now with that moved to `getContext()`, we need a way to pass the Interceptor instance to `openChannelPsbt()`.
> So I see two approacher here, either move the signal interception out of `getContext()` or make `getContext()` return the `Interceptor` instance.

I think the easiest solutions is to replace `<-signal.ShutdownChannel()` with `<-ctx.Done()`, since we know the context will be canceled when the signal is triggered.

hsjoberg commented:
@halseth Ah I see, thanks!
Fixed and squashed.

halseth reviewed (approved):
Great work, LGTM 👍
